wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java

```diff
@@ -757,6 +751,31 @@ boolean createSnapshotSync(WorkspaceRuntimeImpl runtime, String namespace, Strin
         return devMachineSnapshotFailMessage == null;
     }
 
+    private Optional<String> createSnapshot(MachineImpl machine, String namespace) {
+        try {
+            try {
+                SnapshotImpl oldSnapshot = snapshotDao.getSnapshot(machine.getWorkspaceId(),
+                                                                   machine.getEnvName(),
+                                                                   machine.getConfig().getName());
+                snapshotDao.removeSnapshot(oldSnapshot.getId());
+
+                runtimes.removeSnapshot(oldSnapshot);
+            } catch (NotFoundException ignored) {
+                // Do nothing if no snapshot found
+            }
+
+            SnapshotImpl snapshot = runtimes.saveMachine(namespace,
+                                                         machine.getWorkspaceId(),
+                                                         machine.getId());
+            snapshotDao.saveSnapshot(snapshot);
```

Comment:
If this call fails, then docker still exists, i think you should try to make a synthetic transaction and at least try to remove the image 

e.g.

```
try {
    runtimes.save
    dao.save
} catch {
    try {
         runtimes.remove
    } catch {
         // appropriate logging which says that DOCKER image wasn't removed and due to some error
    }
}
```